### PR TITLE
fix: Added exception message for invalid dependency of a task

### PR DESCRIPTION
**Pull request by Aakash017**
fix: Added exception message for invalid depenedency of a task #230

### CODE REVIEW

File: src/taskgraph/generator.py [hunk 1]
```diff
@@ -328,15 +328,19 @@ def _run(self):
         edges = set()
         for t in full_task_set:
             for depname, dep in t.dependencies.items():
-                edges.add((t.label, dep, depname))
+                if dep not in all_tasks.keys():
+                    raise Exception(
+                        f" Task '{t.label}' lists a dependency that does not exist: '{dep}'"
+                    )
+                else:
```
Comment (ahal) [inline src/taskgraph/generator.py:335]: nitpick: Please remove this `else` clause as there's no need for it.

File: taskcluster/ci/doc/kind.yml [hunk 2]
```diff
@@ -25,6 +25,8 @@ task-defaults:
 
 tasks:
     generate:
+#        dependencies:
+#            build: build-linux
```
Comment (ahal) [inline taskcluster/ci/doc/kind.yml:29]: Please remove this commented out config.

File: .gitignore [hunk 3]
```diff
@@ -22,3 +22,6 @@ __pycache__/
 
 # Artifacts
 artifacts/
+
+#Idea
+.idea
```
Comment (ahal) [inline .gitignore:27]: Thanks!

**Review by ahal — CHANGES_REQUESTED** [review]
Excellent, this is exactly what I was hoping for :). I'm requesting changes for the minor issues in comments below.

I also have two optional improvements that will make this a bit more challenging. But this is a clear improvement on the status quo, so if you'd like to land this without these improvements, that's fine with me!

Optional improvements:
1. Add a test that covers this scenario. You can copy/paste [this test](https://github.com/taskcluster/taskgraph/blob/d339ac59186fc24fa8afb4162ada1b5fac44d2a9/test/test_generator.py#L39) as a starting point, and use the `pytest.raises` assertion function to assert an exception is raised.
2. Refactor this so it detects all invalid dependencies for all tasks and prints every single one of them at once (this way we avoid the scenario where someone fixes one broken dependency, only to hit the next one).

File: src/taskgraph/generator.py (hunk 1 above)
Comment (Aakash017) [reply src/taskgraph/generator.py:335]:
```
Task 'doc-generate' lists a dependency that does not exist: 'build-linux'
Traceback (most recent call last):
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 828, in main
    args.command(vars(args))
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 411, in show_taskgraph
    generate_taskgraph(options, parameters, logdir)
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 190, in generate_taskgraph
    out = format_taskgraph(options, spec, logfile(spec))
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 146, in format_taskgraph
    tg = getattr(tgg, options["graph_attr"])
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 171, in full_task_graph
    return self._run_until("full_task_graph")
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 424, in _run_until
    k, v = next(self._run)
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 337, in _run
    raise Exception("Tasks have some dependencies that does not exist: ")
Exception: Tasks have some dependencies that does not exist:

```
@ahal for -> Refactor this so it detects all invalid dependencies for all tasks and prints every single one of them at once (this way we avoid the scenario where someone fixes one broken dependency, only to hit the next one)
pls check above if above changes are fine or not

**Comment by Aakash017** [comment]
```
Task 'doc-generate' lists a dependency that does not exist: 'build-linux'
Traceback (most recent call last):
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 828, in main
    args.command(vars(args))
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 411, in show_taskgraph
    generate_taskgraph(options, parameters, logdir)
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 190, in generate_taskgraph
    out = format_taskgraph(options, spec, logfile(spec))
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 146, in format_taskgraph
    tg = getattr(tgg, options["graph_attr"])
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 171, in full_task_graph
    return self._run_until("full_task_graph")
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 424, in _run_until
    k, v = next(self._run)
  File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 337, in _run
    raise Exception("Tasks have some dependencies that does not exist: ")
Exception: Tasks have some dependencies that does not exist:
```
@ahal for -> Refactor this so it detects all invalid dependencies for all tasks and prints every single one of them at once (this way we avoid the scenario where someone fixes one broken dependency, only to hit the next one)
pls check above if above changes are fine or not

File: src/taskgraph/generator.py (hunk 1 above)
Comment (ahal) [reply src/taskgraph/generator.py:335]: Yeah the message looks good though, you'll need to actually list the dependencies afterwards. I'd think something like:
```
Exception: Some tasks have dependencies that do not exist!

  doc-generate depends on:
    - build-bar
    - build-foo

  build-linux depends on:
    - invalid-task
```

**Comment by ahal** [comment]
> ```
> Task 'doc-generate' lists a dependency that does not exist: 'build-linux'
> Traceback (most recent call last):
>   File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 828, in main
>     args.command(vars(args))
>   File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 411, in show_taskgraph
>     generate_taskgraph(options, parameters, logdir)
>   File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 190, in generate_taskgraph
>     out = format_taskgraph(options, spec, logfile(spec))
>   File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/main.py", line 146, in format_taskgraph
>     tg = getattr(tgg, options["graph_attr"])
>   File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 171, in full_task_graph
>     return self._run_until("full_task_graph")
>   File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 424, in _run_until
>     k, v = next(self._run)
>   File "/Users/aakashsharma/Documents/projects/taskgraph/src/taskgraph/generator.py", line 337, in _run
>     raise Exception("Tasks have some dependencies that does not exist: ")
> Exception: Tasks have some dependencies that does not exist:
> ```
> 
> @ahal for -> Refactor this so it detects all invalid dependencies for all tasks and prints every single one of them at once (this way we avoid the scenario where someone fixes one broken dependency, only to hit the next one) pls check above if above changes are fine or not

Replied in-line. It looks like no changes were pushed to this PR since my initial review, so unsure what you mean by "above changes" :)

**Comment by Aakash017** [comment]
@ahal 

```
 |     [Exception(f"task {t} depends on: \n -- {d}") for t, d in invalid_dependencies.items()]
  | exceptiongroup.ExceptionGroup: Some tasks have dependencies that do not exist! (1 sub-exception)
  +-+---------------- 1 ----------------
    | Exception: task doc-generate depends on: 
    |  -- ['build-linux']
    +------------------------------------
```

Using ExceptionGroup we can raise mutliple exceptiosn but we have to add in requirements folder so pls suggest if i make those changes or look for any other solution

**Comment by Aakash017** [comment]
@ahal changes are done as it were earlier,

**Review by ahal — APPROVED** [review]
Thanks!

I took the liberty of re-adding some whitespace that looks like it was accidentally removed.

**Pull request merged**